### DISCUSSION TITLE
Padrões de construção de API's

### DISCUSSION THREAD

**Discussion by ThiagoBussola** [TI-UNICESUMAR/2024-desafio-profissional-v-ESOFT5S-A#1] [Web Development]
Em sala de aula vimos um pouco sobre o que são API's e o padrão REST.

Faça uma breve pesquisa e escreva como devem ser construidas API's REST e também pesquise outros padrões de construção de API's, quais suas diferenças para o padrão REST e quais suas vantagens e cenários de uso.

**Comment by joaoagr1** [comment]
Algumas coisas que eu vi sobre o padrão rest:
-separação de responsabilidades de interface usuario e banco de dados;
-o conceito de stateless que significa que cada requisição HTTP enviada pelo cliente contém todas as informações necessárias para que o servidor compreenda e processe a requisição, sem depender de nenhum estado armazenado no servidor entre as requisições. então por exemplo quando temos um token de autenticação esse token deve estar presente em todas as requisições da aplicação pois não existe o conceito de "autenticado" ou "nao autenticado" que nem na arquitetura MVC apoiada pelo Spring Security (usando Java como exemplo);

Outros padrões de construção que eu vi são o GraphQL e o gRPC

**Comment by RicardooFilho** [comment]
Quando uma API segue os padrões definidos pela abordagem REST, ela é considerada uma API RESTful. Aplicações RESTful seguem padrões para comunicar entre si, a partir de requisições HTTP, e para enviar estas mensagens existem padrões de formatação que são utilizados. Um deles seria o JSON, que é atualmente o mais utilizado para realizar esta comunicação.
Existem algumas boas práticas na construção de uma API RESTful, visto que seu uso é útil não somente para proporcionar um melhor funcionamento do sistema, mas também para facilitar o entendimento de outras pessoas que venham a utilizar/acessar a aplicação e aplicar uma manutenção no futuro, quando necessário.

- Os recursos da API devem ser representados por substantivos no plural, e não em verbos.

		localhost:8080/pessoas
		

- Para acessar um recurso específico, utilizar o identificador (@Id) como endpoint:

		localhost:8080/pessoas/1
		

- Caso exista um encadeamento de posts, em que tenham recursos atrelados a outros recursos e/ou identificadores, é recomendado manter no máximo um duplo encadeamento de endpoints.

		localhost:8080/pessoas/1/enderecos/2

Além disso, ao falar sobre boas práticas em APIs RESTful é importante mencionar os verbos HTTP, em que os mais utilizados são GET, POST, PUT e DELETE. Estes são utilizados diretamente junto aos URIs, então quando usamos um verbo GET junto a uma URI "/pessoas", já temos ciência que se trata de um método que irá buscar dados de pessoas. Junto dos verbos HTTP, existem os códigos usados para representar o sucesso, erro ou outro resultado proveniente da requisição enviada. Durante o tráfego das requisições e suas respostas, é utilizado o modelo JSON para formatar as mensagens. Os códigos mais utilizados para enviar respostas referentes ao métodos HTTP são:

- GET

		 	200: Sucesso
			404: Entidade não encontrada

- POST

			200: Processo executado, mas sem criar recurso
			201: Recurso criado com sucesso
			400: Requisição com dados inválidos (Bad Request)

- PUT

			200: Recurso atualizado
			201: Recurso inexistente criado
			409: Conflito na atualização do recurso

- DELETE

			204: Sucesso
			404: Entidade não encontrada

Outros modelos de construção de API's que encontrei foram o Falcor, desenvolvido pela Netflix, e o SOAP. As principais diferenças desses dois para o padrão REST estão na abordadem de design, forma como eles tratam as comunicações entre sistemas e forma de recuperação e manipulação de dados.

**Comment by ashura-senpai** [comment]
**O que eu achei sobre APIs REST é que elas são projetadas seguindo princípios específicos que existem pra facilitar a comunicação entre sistemas. Alguns exemplos de diretrizes para construir APIs REST:**

1. Utilização de Métodos HTTP: As operações CRUD (Create, Read, Update, Delete) são mapeadas nos métodos HTTP correspondentes (POST, GET, PUT, DELETE).

2. URI (Identificador de Recurso Uniforme): As URIs devem ser utilizadas para identificar recursos de maneira única. Cada recurso deve ter uma URI única.

3. Representação dos Recursos: Os recursos são representados geralmente em formato JSON ou XML. O formato escolhido pode ser negociado entre o cliente e o servidor.

4. Stateless (Sem Estado): Cada requisição do cliente para o servidor contém todas as informações necessárias para entender e processar a requisição. Não há necessidade de manter um estado de sessão no servidor.

5. HATEOAS (Hypermedia As The Engine Of Application State): Os clientes interagem com a aplicação inteiramente por meio de hiperlinks fornecidos de maneira dinâmica pelos aplicativos servidores.

6. Uso de Códigos de Status HTTP: Os códigos de status HTTP são utilizados para indicar o sucesso ou falha de uma requisição.

**Achei também outros padrões de Construção de APIs, porém, cada padrão tem suas vantagens e é mais adequado para determinados cenários. A escolha entre eles depende dos requisitos específicos do projeto, como eficiência, simplicidade, flexibilidade, etc. Exemplos abaixo:**

1. GraphQL:
   - Diferenças com REST: Enquanto o REST utiliza múltiplos endpoints para diferentes recursos, o GraphQL oferece uma única entrada para receber ou enviar dados.
   - Vantagens: Permite que o cliente especifique exatamente quais dados ele precisa, evitando over-fetching ou under-fetching de dados.
   - Cenários de Uso: Especialmente útil em casos onde a eficiência na transferência de dados é crucial.

2. SOAP (Simple Object Access Protocol):
   - Diferenças com REST: Usa XML para a representação dos dados e opera principalmente sobre HTTP, SMTP ou outros protocolos.
   - Vantagens: Rigoroso e altamente extensível, com suporte a transações e segurança embutidos.
   - Cenários de Uso: Comum em ambientes corporativos onde integridade e segurança são críticas.

3. gRPC (gRPC Remote Procedure Calls):
   - Diferenças com REST: Usa protocolo binário, é mais eficiente em termos de desempenho e suporta bidirecional streaming.
   - Vantagens: Eficiência, suporte a várias linguagens de programação, e streaming eficiente.
   - Cenários de Uso: Comunicação eficiente em ambientes distribuídos e em sistemas onde a baixa latência é crucial.
   
  É isso, demorei mas foi. :)

**Comment by EnzoLavieri** [comment]
As APIs REST (Representational State Transfer) é uma abordagem arquitetônica para projetar serviços web. E eu separei nos tópicos  abaixo, de como as APIs REST devem ser construídas:

- Use Métodos HTTP Adequados:
   - Utilize os métodos HTTP de maneira apropriada para cada operação: 
      **GET** para leitura;
      **POST** para criação;
      **PUT/PATCH** para atualização; 
      **DELETE** para exclusão.
   - Evite o uso de métodos não seguros, como GET, para as operações que modificam dados.

- URLs Navegáveis e Intuitivas:
   - Mantenha URLs intuitivas e representativas dos recursos.
   - Utilize substantivos para representar recursos em vez de verbos.

- Padrões de Nomenclatura:
   - Use nomes de recursos no plural (por exemplo, `/usuarios` em vez de `/usuario`).
   - Utilize camelCase ou snake_case para nomes de campos em JSON.

- Versionamento da API:
   - Inclua uma versão na URL da API (por exemplo, `/v1/usuarios`) para garantir compatibilidade futura.

- Tratamento de Erros Consistente:
   - Forneça mensagens de erro claras e consistentes usando códigos de status HTTP apropriados.

- Autenticação e Autorização:
   - Utilize protocolos de autenticação padrão, como OAuth, para proteger a API.
   - Implemente mecanismos de autorização para controlar o acesso aos recursos.

- Respostas Estruturadas:
   - Utilize JSON como formato padrão para respostas.
   - Inclua links relacionados, HATEOAS, para facilitar a navegação na API.

- Tratamento de Relacionamentos:
   - Ao lidar com relacionamentos entre recursos, forneça mecanismos para inclusão de dados relacionados nas respostas (por exemplo, usando a técnica de "embedding" ou "linking").

- **Logs e Monitoramento:**
    - Registre eventos importantes para facilitar a solução de problemas e o monitoramento da saúde da API.

- Documentação Clara:
    - **FORNEÇA UMA DOCUMENTAÇÃO** clara e abrangente, incluindo exemplos de solicitações e respostas.

- Testes:
    - Desenvolva testes automatizados ou manuais para garantir a estabilidade, confiabilidade e usabilidade da API.
    
 Além das diretrizes REST, há outros padrões e estilos de construção de APIs. Abaixo irei mencionar esses outros tipos de padrões:

- GraphQL:
   - **Diferenças:** Ao contrário do REST, onde o cliente tem pouca influência sobre a estrutura da resposta, o GraphQL permite que os clientes especifiquem os dados exatos de que precisam. As consultas são definidas pelos clientes, o que reduz a quantidade de dados transferidos.
   - **Vantagens:**
      - Redução do over-fetching (transferência de dados desnecessários).
      - Permite aos clientes buscar dados de várias fontes em uma única consulta.
   - **Cenários de Uso:** Útil em casos onde a eficiência na transferência de dados é crítica e em ambientes com muitos clientes diferentes que precisam de visualizações específicas dos dados.

- gRPC (gRPC Remote Procedure Calls):
   - **Diferenças:** Usa o protocolo HTTP/2 para comunicação binária e define serviços com contratos de interface definidos usando Protocol Buffers (protobuf). Utiliza RPC (Remote Procedure Call) para a comunicação entre cliente e servidor.
   - **Vantagens:**
      - Eficiência na serialização binária e na transmissão de dados.
      - Geração automática de código a partir dos contratos de serviço.
   - **Cenários de Uso:** Ambientes onde a eficiência na comunicação é crucial, como em sistemas distribuídos e comunicações entre microsserviços.

- SOAP (Simple Object Access Protocol):
   - **Diferenças:** Utiliza XML para a formatação da mensagem e geralmente opera sobre o protocolo HTTP, embora não esteja limitado a ele. Define operações utilizando um contrato WSDL (Web Services Description Language).
   - **Vantagens:**
      - Pode oferecer funcionalidades robustas de segurança.
      - Suporte a transações e compensações.
   - **Cenários de Uso:** Ambientes corporativos que necessitam de recursos avançados de segurança, transações e controle de versão.

- **Falcor:**
   - **Diferenças:** Desenvolvido pela Netflix, o Falcor permite que os clientes especifiquem os dados necessários usando um modelo de dados JSON referenciado.
   - **Vantagens:**
      - Redução do under-fetching (quando dados insuficientes são recuperados).
      - Flexibilidade na definição de consultas.
   - **Cenários de Uso:** Aplicações que requerem otimização no carregamento de dados e que necessitam de um modelo de dados flexível.

Existe outros padrões de construção de API's, como JSON-RPC, XML-RPC, OData, e mais alguns, só não coloquei se não vai ficar mais longo do que já está :p
